### PR TITLE
[Deepseek] Support absorbed-MLA attention 

### PR DESCRIPTION
**Pull request by JamesBrianD**
## Motivation

Land the absorbed MLA path end-to-end (see `docs/design/MLA.md`): kernel + backend + KV-pool plumbing, model-side absorbed `__call__`, model-runner dispatch, and a numerical-correctness test against the v1 reference. Also expose `fa_mha` as a non-absorbed backend choice so users can A/B against the absorbed path on the same MLA model.

## Modifications

**Kernel — vendor + sglang-jax adaptation** (`python/sgl_jax/srt/kernels/mla/v2/kernel.py`)
- Vendored tpu-inference's MLA v2 Pallas kernel (Apache-2.0 header preserved + `Adapted from` line added).
- Adapted the KV-cache addressing to sglang-jax's ragged `page_indices` layout (each seq's pages are tightly concatenated, **not** padded to a uniform `pages_per_seq`).
  - Added `cu_kv_lens` as a scalar prefetch on the inner kernel and as an outer-API parameter (passed through `static_validate_inputs`, `run_mla_kernel`, and all 3 call sites).
  - Replaced the padded-layout offset `seq_idx * pages_per_seq + kv_p_start` (fetch + cache-write paths) with `cdiv(cu_kv_lens_ref[seq_idx], page_size) + kv_p_start`, matching RPA v3.
  - Dropped the `num_page_indices % max_num_seqs == 0` assertion (no longer holds under ragged layout).
  - Bumped `input_output_aliases` indices `{7:0, 11:1} → {8:0, 12:1}` to account for the extra scalar prefetch.

**Reference implementation** (`python/sgl_jax/srt/kernels/mla/v1/ref.py`)
- Vendored only `ref_mla_ragged_paged_attention` (and its direct deps: `update_kv_cache`, `static_validate_inputs`, `dynamic_validate_inputs`, `get_kv_cache_shape`) from tpu-inference's MLA v1 kernel — the full v1 Pallas kernel is intentionally not ported. This reference exists purely for numerical-correctness checks against v2.

**Backend — absorbed-MLA attention** (`python/sgl_jax/srt/layers/attention/mla_backend.py`)
- New `MLAAttentionBackend` + `MLAAttentionMetadata`, contract matching the other attention backends so `RadixAttention` stays the unified entry point: `(q, k, v, layer, forward_batch, pool, **kwargs)` with `q_rope` / `k_rope` in kwargs. The backend internally reassembles the 4-tuple `(ql_nope, q_pe, new_kv_c, new_k_pe)` the v2 Pallas kernel consumes, and returns `(o_latent [T, n_h, kv_lora_rank], updated_cache)` — caller projects via `W_UV → W_O`.
- `get_forward_metadata` builds `cu_q_lens` / `cu_kv_lens` / `page_indices` / `seq_lens` / `distribution` for both EXTEND and DECODE; EXTEND routes through the mixed branch (`distribution = [0, 0, num_seqs]`) since the v2 kernel skips the chunked-prefill-only branch.
- Pipes `cu_kv_lens` through the `shard_map` to the kernel.

**KV pool — 4D paged latent layout** (`python/sgl_jax/srt/mem_cache/memory_pool.py`)
- `MLATokenToKVPool` reshaped to the 4D paged buffer the kernel expects: `[total_num_pages, page_size_per_kv_packing, kv_packing, align_to(kv_lora_rank, 128) + align_to(qk_rope_head_dim, 128)]`. The two segments are aligned **independently** — see the docstring for the kernel-ABI rationale.

**Model — absorbed forward + MHA forward** (`python/sgl_jax/srt/models/deepseek_v3.py`)
- Pre-compute `W_UK` / `W_UV` placeholders (`nnx.Param`, sharded `(None, "tensor", None)`) at `__init__`; `post_load_weights()` reshapes `kv_b_proj.weight` into `(kv_lora_rank, num_heads, qk_nope+v_head_dim)`, slices into `w_uk`/`w_uv`, and drops `kv_b_proj` (mirrors sglang's `deepseek_weight_loader.post_process()`).
- `_forward_mqa` (absorbed): fold `W_UK` into Q (`einsum("thd,rhd->thr", q_nope, w_uk)`), pack latent `c_kv` as `[T, 1, kv_lora_rank]`, call `attn_mqa`, then fold `W_UV` into the output before `o_proj`.
- `_forward_mha` (non-absorbed): decompress latent via `kv_b_proj` to per-head K/V, concat `[k_nope, k_rope]`, pad V on the head dim (RPA v3 requires k/v share head_dim), call `attn_mha`, strip V padding from the output.
- `__call__` dispatches on `self.use_absorbed`, set from `hf_config.use_absorbed_mla` at `__init__`.

**Model runner — backend & pool dispatch** (`python/sgl_jax/srt/model_executor/model_runner.py`)
- `ServerArgs.attention_backend` choices: `native | fa | fa_mha`. `fa` selects absorbed (MLA Pallas kernel + MLA pool); `fa_mha` and `native` select non-absorbed (FlashAttention / NativeAttention + MHA pool).
- Patches `hf_config.use_absorbed_mla = (attention_backend == "fa")` once at server start, so the model can read it during `__init__`.
- KV pool: MLA pool only when MLA model AND `fa`; otherwise MHA pool (covers non-MLA models, plus `fa_mha`/`native` on MLA models).
- `_compute_cell_size` (drives `profile_max_num_token`):
  - absorbed MLA: `(align128(kv_lora_rank) + align128(qk_rope_head_dim)) * num_layers * dtype` (latent c_kv only, no `*2`; pool is replicated, so per-device == total).
  - everything else: `get_num_kv_heads(tp_size) * align128(model_config.head_dim) * 2 * num_layers * dtype` (per-device, MHA pool sharded on the head dim). For MLA models, `patch_model_config` sets `head_dim = qk_nope+qk_rope` so the same formula works.
- `_forward` device_put now reads `self.token_to_kv_pool.kv_sharding` instead of a hardcoded 5D PartitionSpec, so the 4D MLA pool and the 5D MHA pool both work without branching.

**Test — MLA backend correctness** (`python/sgl_jax/test/test_mla_attention.py`, registered in `test/srt/run_suite.py:unit-test-tpu-v6e-1`)
- New test mirroring `test_flashattention.py`, validating `MLAAttentionBackend` outputs against `ref_mla_ragged_paged_attention` across EXTEND / DECODE modes.

**E2E test — DeepSeek-Coder-V2-Lite-Instruct GSM8K** (`test/srt/test_deepseek_v2_lite_models.py`, registered in `test/srt/run_suite.py:e2e-test-tpu-v6e-4`)
- New end-to-end test launching the server with the absorbed-MLA backend on `deepseek-ai/DeepSeek-Coder-V2-Lite-Instruct` and running 200-question GSM8K via `simple_eval_gsm8k` (chat endpoint). Threshold `0.85` matches sglang's nightly `test_text_models_gsm8k_eval.py`.

**Misc**
- `python/pyproject.toml`: added MLA v2 kernel to `[tool.ruff.lint.per-file-ignores]` for `B023` (Pallas for-loops are statically unrolled at trace time).

## Backend selection on MLA models — `fa` (absorbed) vs `fa_mha` (non-absorbed)

Both paths share the front half of `DeepseekV3Attention.__call__` (Q projection, KV-A latent + RoPE). They diverge in **what gets cached** and **how attention is computed**:

| | `fa` (absorbed MLA) | `fa_mha` (non-absorbed) |
|---|---|---|
| Forward branch | `_forward_mqa` | `_forward_mha` |
| `kv_b_proj` | dropped after `post_load_weights` (folded into `w_uk`/`w_uv`) | kept; runs every forward to decompress latent → per-head K/V |
| Q transformation | `ql_nope = einsum("thd,rhd->thr", q_nope, w_uk)` (fold UK into Q) | `q = concat([q_nope, q_rope], -1)` |
| K used by attention | `[c_kv, k_rope]` — single shared latent head, MQA-shaped `[T, 1, *]` | per-head `[k_nope ⊕ k_rope]`, shape `[T, n_h, qk_head_dim]` |
| V used by attention | latent `c_kv` (same as K) | per-head V, padded to `qk_head_dim` (RPA v3 requires k/v share head_dim) |
| Attention kernel | MLA v2 Pallas kernel via `MLAAttentionBackend` | RPA v3 via `FlashAttention` |
| Output post-processing | `einsum("thr,rhd->thd", o_latent, w_uv)` (fold UV into output) | strip V padding from `o[..., :v_head_dim]` |
| KV pool | `MLATokenToKVPool` — 4D `[pages, page_size//kv_packing, kv_packing, align128(lora)+align128(rope)]`, replicated across TP | `MHATokenToKVPool` — 5D `[pages, page_size, n_kv*2//kv_packing, kv_packing, align128(qk_head_dim)]`, sharded on head dim |
| Per-token KV bytes (DS-V3, bf16) | `(align128(512)+align128(64)) * 61 * 2 = ~76 KB` (per device, replicated → same on every device) | `(128/tp) * align128(192) * 2 * 61 * 2` per device — at tp=4: `32 * 256 * 2 * 61 * 2 ≈ 2 MB` per device |
| Compute trade-off | Pallas kernel has its own MLA-specific schedule; cache reads/writes are the small latent | Reuses standard FA path; extra `kv_b_proj` matmul + much larger cache I/O each step |

When to pick which:
- **`fa` (default for MLA models)**: long-context inference. Absorbed cache is ~25–30× smaller per token, which is the whole point of MLA; expect this to win once batch × context is non-trivial.
- **`fa_mha`**: A/B / debugging the absorbed path against the standard FlashAttention path, or short-context experiments where pool size isn't the bottleneck. Auto-falls back to the same MHA pool plumbing the rest of the project already exercises.
- **`native`**: same non-absorbed forward path as `fa_mha`, but routes through `NativeAttention` instead of FA. Used by the existing native-attention CI lane.

## Test plan

```bash
python -m pytest python/sgl_jax/test/test_mla_attention.py -v
```

## Accuracy

End-to-end accuracy validated on **DeepSeek-Coder-V2-Lite-Instruct** with the absorbed-MLA path (`fa` backend) on **TPU v6e (2x2, tp=4)**, matching sglang's nightly setup (`test/manual/nightly/test_text_models_gsm8k_eval.py`).

**Server launch:**

```bash
python -u -m sgl_jax.launch_server \
  --model-path deepseek-ai/DeepSeek-Coder-V2-Lite-Instruct --trust-remote-code \
  --dist-init-addr=0.0.0.0:10011 --nnodes=1 --tp-size=4 \
  --device=tpu --node-rank=0 \
  --mem-fraction-static=0.8 --max-prefill-tokens=8192 \
  --dtype=bfloat16 --skip-server-warmup \
  --host 0.0.0.0 --port 30000 \
  --page-size 128 --context-length 8192
```

> Note: `--attention-backend` accepts `native | fa | fa_mha`. For MLA models, `fa` selects the absorbed path; `fa_mha` and `native` select the non-absorbed (MHA) path. `--page-size > 1` is required for the absorbed path — see `MLAAttentionBackend.get_max_running_reqests` (asserts `res > 0`).

**GSM8K (chat-template eval via `simple_eval_gsm8k` / `/v1/chat/completions`):**

| # questions | Score | Eval latency | Threshold | Notes |
|---|---|---|---|---|
| 200 | **0.880** | 59.7 s | — | smoke test |
| **1319 (full)** | **0.859** | 316.9 s | **0.85** ✅ | matches sglang nightly threshold |

The 1319-question full run ties sglang's official nightly bar (0.85) for this model, confirming the absorbed `W_UK`/`W_UV` einsum + Pallas v2 kernel is numerically equivalent to the reference path end-to-end.

**Earlier base-model run** (`DeepSeek-V2-Lite` base + raw `/generate` few-shot via `bench_sglang_jax.py`, kept for reference; chat-template eval above is the canonical one):

| Metric | Value |
|---|---|
| Accuracy | 0.378 |
| Invalid | 0.006 |
| Latency | 150.698 s |
| Output throughput | 1061.826 token/s |

## Checklist

- [x] Please use English, otherwise it will be closed.
- [x] The purpose of the PR, or link existing issues this PR will resolve.
- [x] The test plan, such as providing test command.
- [ ] (Optional) The necessary documentation update.

### CODE REVIEW

**Comment by cjx0709** [comment]
`profile_max_num_token` should update for MLA situation, while in sglang `_compute_cell_size`, using `use_mla_backend` for another memory cell size compute logic.

File: python/sgl_jax/srt/models/deepseek_v3.py [hunk 1]
```diff
@@ -232,28 +261,94 @@ def __call__(
...
+        q_nope: jax.Array,
+        q_rope: jax.Array,
+        compressed: jax.Array,
+        k_rope: jax.Array,
+        forward_batch: ForwardBatch,
+        token_to_kv_pool: KVCache,
+    ) -> tuple[jax.Array, jax.Array]:
+        # kv_b_proj.weight is [kv_lora_rank, n_h * (qk_nope + v_head_dim)],
+        # head-major with [nope, v] within each head block.
+        w_kv = self.kv_b_proj.weight.value.reshape(
```
Comment (cjx0709) [inline python/sgl_jax/srt/models/deepseek_v3.py]: this reshape and later split maybe can do this in weight load

Comment (JamesBrianD) [reply python/sgl_jax/srt/models/deepseek_v3.py]: Add `post_load_weights` to fix.

**Comment by JamesBrianD** [comment]
> `profile_max_num_token` should update for MLA situation, while in sglang `_compute_cell_size`, using `use_mla_backend` for another memory cell size compute logic.

Thanks for suggestion. Refactor the `profile_max_num_token` to get the accurate cell size.

File: python/sgl_jax/srt/model_executor/model_runner.py
Comment (cjx0709) [inline python/sgl_jax/srt/model_executor/model_runner.py:753]: this is 5D while MLA mempool shape is 4D

File: python/sgl_jax/srt/model_executor/model_runner.py [hunk 2]
```diff
@@ -306,6 +306,13 @@ def load_model(self):
             self.server_args.ep_num_redundant_experts
         )
         self.model_config.hf_config.moe_backend = self.model_config.moe_backend.value
+        # Pick MLA forward path at server start: `fa` → absorbed Pallas kernel
+        # path; `fa_mha` → non-absorbed FlashAttention path. Read by
+        # DeepseekV3DecoderLayer to construct DeepseekV3Attention; harmless on
+        # non-MLA models that ignore the attribute.
+        self.model_config.hf_config.use_absorbed_mla = (
+            self.server_args.attention_backend != "fa_mha"
```
Comment (cjx0709) [inline python/sgl_jax/srt/model_executor/model_runner.py]: if backendend == native should be no-absorbed

**Comment by aolemila** [comment]
0.378 is a little low compared with 0.41. Could you evaluate another two times to get the average grade?

File: python/sgl_jax/srt/mem_cache/memory_pool.py [hunk 3]
```diff
@@ -874,7 +874,19 @@ def _sharded_update(local_fused_kv, local_loc, local_kv_cache):
     return _sharded_update(fused_kv, loc, kv_cache)
 
 
+@register_pytree_node_class
 class MLATokenToKVPool(KVCache):
+    """Paged KV cache for absorbed-MLA path.
+
+    Layout matches the MLA v2 Pallas kernel (`get_kv_cache_shape`):
+      `[num_pages, align_to(page_size, kv_packing)//kv_packing, kv_packing,
+        align_to(kv_lora_rank, 128) + align_to(qk_rope_head_dim, 128)]`
```
Comment (aolemila) [inline python/sgl_jax/srt/mem_cache/memory_pool.py:883]: It is different from `align_to(kv_lora_rank + qk_rope_head_dim, 128)` which is recorded in PR description. For the corner case whose kv_lora_rank is 513 and qk_rope_head_dim is 64, they get the different results. 
`align_to(kv_lora_rank + qk_rope_head_dim, 128) = 5, align_to(kv_lora_rank, 128) + align_to(qk_rope_head_dim, 128) = 5+1=6`

File: python/sgl_jax/srt/model_executor/model_runner.py
Comment (JamesBrianD) [reply python/sgl_jax/srt/model_executor/model_runner.py:753]: Fixed

**Comment by JamesBrianD** [comment]
DeepSeek-Coder-V2-Lite-Instruct achieved a test score of 85.9 on the `gsm8k` dataset, consistent with the score on HF.

**Review by aolemila — APPROVED** [review]
